### ISSUE TITLE
Non-deterministic problem building

### ISSUE THREAD

**Issue by joaquimg** [spine-tools/SpineOpt.jl#1143]
I am running the `main.jl` script in this PR https://github.com/jump-dev/open-energy-modeling-benchmarks/pull/38

Every time I run it, I get a different mps file.

Rows are certainly and non-deterministic order:
![image](https://github.com/user-attachments/assets/dbfafb19-107e-418e-9d03-10b257939fab)

This might be due to many reasons:
- Some bug
- Iterating in the Julia `Dict` (or some other collection that does not preserve ordering). `OrderedDict` from `OrderedCollection` is slower for some operations but preserves the order.

Even if the only issue is order, the models might have heavily varying solution times and might be difficult to debug.

Deterministic behavior would be ideal here.

**Comment by Tasqu** [comment]
There are a couple of places I think might cause this:
1. Our use of `Dict`s in storing JuMP variable and constraint references, as well as other stuff like temporal and stochastic structure information. Essentially anything we store in `m.ext[:spineopt]`. However, I'm not sure if this stuff impacts the JuMP model `m` generation directly.
2. Could potentially be caused by the indexing functions we use? I'm not sure if the order of entities is consistent when it is fetched from the database.

**Comment by DillonJ** [comment]
@joaquimg I also believe that these days we use parallel processing to build constraints, so I expect that can also cause changes in ordering etc. Can you confirm @manuelma ? If this potentially causes some variation in model solve time, is there any reason to expect that there is any net negative impact?

@joaquimg are you aware of the [write_model_file()](https://spine-tools.github.io/SpineOpt.jl/latest/how_to/print_the_model/) function? This is likely a much better method for debugging models as you get constraint and variable names.

**Comment by manuelma** [comment]
I'm probably missing something but quickly looking at the model json file I found that mip_rel_gap is set to 0.50 which maybe means 50% gap? Could that be causing the difference in solution times?

@DillonJ I don't see any reason why multithreading in building the constraints could lead to any issue regarding solve.

**Comment by DillonJ** [comment]
@manuelma @joaquimg is just talking about the model formulation... nothing about solving the model. When the mps file is written, it is not identical for the same problem

**Comment by manuelma** [comment]
@DillonJ right, I got that but thought you were asking about model solve time. So mutlithreading - as far as I can see - could be responsible for constraints being generated and possibly written to the mps in random order, but still it will always be the same model and thus solve time shouldn't be affected. I don't know what the solver does but I could imagine that the order of the rows in a matrix could result in differences in how difficult it is to fatorize it? No idea, just wild guessing...

**Comment by joaquimg** [comment]
Perfect. What I said about `@expression` is valid for `@build_constraint` as well. Thanks.
